pinot-broker/src/main/java/org/apache/pinot/broker/routing/builder/BalancedRandomRoutingTableBuilder.java

```diff
@@ -43,7 +43,8 @@
   public void init(Configuration configuration, TableConfig tableConfig, ZkHelixPropertyStore<ZNRecord> propertyStore,
       BrokerMetrics brokerMetrics) {
     super.init(configuration, tableConfig, propertyStore, brokerMetrics);
-    _numRoutingTables = configuration.getInt(NUM_ROUTING_TABLES_KEY, DEFAULT_NUM_ROUTING_TABLES);
+    int numReplicas = tableConfig.getValidationConfig().getReplicationNumber();
+    _numRoutingTables = numReplicas == 1 ? 1 : configuration.getInt(NUM_ROUTING_TABLES_KEY, DEFAULT_NUM_ROUTING_TABLES);
```

Comment:
where is this configuration coming from? My main concern about this change is if the replicationFactor changes and we invoke rebalance, the number of routingTable remains the same

Reply:
The replication is coming from tableConfig -- segmentsConfig part of tableConfig.

```
"segmentsConfig": {
      "completionConfig": null, 
      "hllConfig": null, 
      "replicaGroupStrategyConfig": {
        "numInstancesPerPartition": 2, 
        "partitionColumn": null
      }, 
      "replicasPerPartition": null, 
      **"replication": "3",** 
      "retentionTimeUnit": "DAYS", 
      "retentionTimeValue": "7320", 
    }

If the replication changes, the change should be reflected in table config right? So on the next change to external view, the routing table builder should see the updated config and thus the new replication factor. Is my understanding correct?

Reply:
Just FYI.... we had seen this in couple of non prod deployments where there were 1000s of segments across 20+ servers but replication 1 and the routing table looked weird with duplicate server -> segments mapping

Reply:
If we are not guaranteed to see the updated table config after the replication factor changes then yes we should not be doing this change. AFAIK, there is no way to change the replication factor without going via table config